third_party/move/move-vm/runtime/src/native_layout_cache.rs

```diff
@@ -0,0 +1,87 @@
+
+/// Cache for all layouts available for native context. When VM calls into a native function, these
+/// layouts are the only layouts accessible. The Move implementation must ensure layouts are pre-
+/// loaded into cache before the native call.
+#[derive(Default)]
+pub(crate) struct NativeLayoutCache {
+    /// Runtime layouts that are allowed to be used by natives.
+    layouts: BTreeMap<Type, LayoutWithDelayedFields>,
+    /// Annotated layouts that are allowed to be used by natives (e.g., `string_utils.move`).
+    annotated_layouts: BTreeMap<Type, LayoutWithDelayedFields>,
+}
+
+impl NativeLayoutCache {
+    /// Returns the runtime layout from the cache. If it does not exist, an invariant violation is
+    /// returned.
+    pub(crate) fn type_to_type_layout_with_delayed_fields(
+        &self,
+        ty: &Type,
+    ) -> PartialVMResult<LayoutWithDelayedFields> {
+        self.layouts.get(ty).cloned().ok_or_else(|| {
+            PartialVMError::new_invariant_violation(format!("Layout for {:?} is not cached!", ty))
+        })
+    }
+
+    /// Returns the annotated layout from the cache. If it does not exist, an invariant violation
+    /// is returned.
+    pub(crate) fn type_to_annotated_type_layout_with_delayed_fields(
+        &self,
+        ty: &Type,
+    ) -> PartialVMResult<LayoutWithDelayedFields> {
+        self.annotated_layouts.get(ty).cloned().ok_or_else(|| {
+            PartialVMError::new_invariant_violation(format!(
+                "Annotated layout for {:?} is not cached!",
+                ty
+            ))
+        })
+    }
+
+    /// Converts a vector if types to layouts, and adds them in the cache if they are not there
+    /// yet.
+    pub(crate) fn insert(
+        &mut self,
+        layout_converter: &LayoutConverter<impl StructDefinitionLoader>,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut TraversalContext,
+        tys: Vec<Type>,
+        annotated: bool,
+    ) -> PartialVMResult<()> {
+        for ty in tys {
+            #[allow(clippy::collapsible_else_if)]
+            if annotated {
+                if let Entry::Vacant(entry) = self.annotated_layouts.entry(ty) {
+                    let layout = layout_converter
+                        .type_to_annotated_type_layout_with_delayed_fields(
+                            gas_meter,
+                            traversal_context,
+                            entry.key(),
+                        )?;
+                    entry.insert(layout);
+                }
+            } else {
+                if let Entry::Vacant(entry) = self.layouts.entry(ty) {
+                    let layout = layout_converter.type_to_type_layout_with_delayed_fields(
+                        gas_meter,
+                        traversal_context,
+                        entry.key(),
+                    )?;
+                    entry.insert(layout);
+                }
+            };
+        }
+        Ok(())
+    }
+}
```

Comment:
if the types involved can be easily mocked, can we have a trivial unit test to check the caching functionality & annotated distinction (i.e. add one, add another one, check expected existence before, in the middle, after).